src/Bound/ScopeT.hs

```diff
@@ -0,0 +1,251 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+#if __GLASGOW_HASKELL__ >= 805
+{-# LANGUAGE QuantifiedConstraints #-}
+#endif
+module Bound.ScopeT (
+    ScopeT (..),
+    (>>>>=),
+    -- * Abstraction
+    abstractT, abstract1T, abstractTEither,
+    -- ** Name
+    abstractTName, abstract1TName,
+    -- * Instantiation
+    instantiateT, instantiate1T, instantiateTEither,
+    -- * Traditional de Bruijn
+    fromScopeT,
+    toScopeT,
+    -- * Bound variable manipulation
+    lowerScopeT,
+    splatT,
+    bindingsT,
+    ) where
+
+import Bound
+import Bound.Name
+import Data.Bifunctor
+import Data.Functor.Classes
+
+-- | @'Scope' b f a@ is a @t f@ expression abstracted over @f@,
+-- with bound variables in @b@, and free variables in @a@.
+--
+-- @'Scope' n f a ~ 'ScopeT' n 'IdentityT' f a@
+-- @'ScopeT' n t f a ~ t ('Scope' n f) a@
+-- 
+newtype ScopeT n t f a = ScopeT { unscopeT :: t f (Var n (f a)) }
+
+-------------------------------------------------------------------------------
+-- Instances
+-------------------------------------------------------------------------------
+
+instance (Functor (t f), Functor f) => Functor (ScopeT n t f) where
+   fmap f (ScopeT a) = ScopeT $ fmap (fmap (fmap f)) a
+
+instance (Foldable (t f), Foldable f) => Foldable (ScopeT n t f) where
+    foldMap f (ScopeT a) = foldMap (foldMap (foldMap f)) a
+    foldr f z (ScopeT a) = foldr (flip (foldr (flip (foldr f))))  z a
+
+instance (Traversable (t f), Traversable f) => Traversable (ScopeT n t f) where
+    traverse f (ScopeT a) = ScopeT <$> traverse (traverse (traverse f)) a
+
+-- | We cannot write @'Bound' ('ScopeT' n t)@ pre-GHC-8.6 (without an auxiliary type class).
+(>>>>=) :: (Monad f, Functor (t f)) => ScopeT n t f a -> (a -> f b) -> ScopeT n t f b
+ScopeT m >>>>= k = ScopeT $ fmap (fmap (>>= k)) m
+{-# INLINE (>>>>=) #-}
+
+#if __GLASGOW_HASKELL__ >= 805
+-- | @(>>>=) :: ... => 'ScopeT' n t f a -> (a -> f b) -> 'ScopeT' n t f b@
+instance (forall f. Functor f => Functor (t f)) => Bound (ScopeT n t) where
```

Comment:
I first went with defining

```haskell
class FunctorT t where
    fmapT :: Functor f => (a -> b) -> t f a -> t f b
```

but it was a rabbit hole, `FoldableT`, `TraversableT`, `EqT`, `OrdT`, `ShowT`, `ReadT`...

So I think, better to admit that `QuantifiedConstraints` are good thing, and until GHC-8.6 people can use `>>>>=`.